File: ios/Policy/TrueImagePolicy.m
#import "TrueImagePolicy.h"

CGRect TrueImageFitRect(CGRect bounds, CGSize intrinsic, TrueImageFitMode mode)
{
  if (intrinsic.width <= 0 || intrinsic.height <= 0 || bounds.size.width <= 0 || bounds.size.height <= 0) {
    return bounds;
  }
  CGSize size;
  switch (mode) {
    case TrueImageFitModeStretch:
      return bounds;
    case TrueImageFitModeCenter:
      size = intrinsic;
      break;
    case TrueImageFitModeCover:
    case TrueImageFitModeContain: {
      CGFloat sx = bounds.size.width / intrinsic.width;
      CGFloat sy = bounds.size.height / intrinsic.height;
      CGFloat scale = mode == TrueImageFitModeCover ? MAX(sx, sy) : MIN(sx, sy);
      size = CGSizeMake(intrinsic.width * scale, intrinsic.height * scale);
      break;
    }
  }
  return CGRectMake(
      CGRectGetMidX(bounds) - size.width / 2, CGRectGetMidY(bounds) - size.height / 2, size.width, size.height);
}

BOOL TrueImageThumbnailPixelSize(
    CGSize imagePixels, CGSize drawnPoints, CGFloat scale, TrueImageFitMode mode, CGSize *outSize)
{
  if (drawnPoints.width <= 0 || drawnPoints.height <= 0 || scale <= 0 || imagePixels.width <= 0 ||
      imagePixels.height <= 0 || mode == TrueImageFitModeCenter) {
    return NO;
  }
  CGRect drawnPixels = CGRectMake(0, 0, drawnPoints.width * scale, drawnPoints.height * scale);
  CGSize target = TrueImageFitRect(drawnPixels, imagePixels, mode).size;
  if (imagePixels.width < target.width * 2 || imagePixels.height < target.height * 2) {
    return NO;
  }
  if (outSize) {
    *outSize = CGSizeMake(ceil(target.width), ceil(target.height));
  }
  return YES;
}

NSString *TrueImageThumbnailKey(NSString *source, CGFloat blurRadius, CGFloat blurDownscale, CGSize pixelSize)
{
  return [NSString stringWithFormat:@"%@|b%g|d%g|%dx%d",
                                    source,
                                    (double)blurRadius,
                                    (double)blurDownscale,
                                    (int)pixelSize.width,
                                    (int)pixelSize.height];
}

CGFloat const TrueImageDefaultBlurPixelsPerRadius = 2;

CGFloat TrueImageBlurDownscaleFactor(CGFloat blurRadius, CGFloat pixelsPerRadius)
{
  if (blurRadius <= 0 || pixelsPerRadius <= 0) {
    return 1;
  }
  return MAX(1, blurRadius / pixelsPerRadius);
}

CGSize TrueImageBlurDownscaleSize(CGSize pixels, CGFloat factor)
{
  if (factor <= 1) {
    return pixels;
  }
  return CGSizeMake(MAX(1, ceil(pixels.width / factor)), MAX(1, ceil(pixels.height / factor)));
}

NSUInteger TrueImageMemoryCacheCost(unsigned long long physicalMemory)
{
  unsigned long long const cap = 256ull * 1024 * 1024;
  return (NSUInteger)MIN(physicalMemory / 16, cap);
}

BOOL TrueImageShouldFade(NSInteger transitionMs, BOOL fromMemory, BOOL hasContents, BOOL isResource)
{
  if (transitionMs <= 0 || isResource) {
    return NO;
  }
  return !(fromMemory && !hasContents);
}

TrueImageFadeKeyPath TrueImageFadeKeyPathFor(BOOL hasContents)
{
  return hasContents ? TrueImageFadeKeyPathContents : TrueImageFadeKeyPathOpacity;
}

NSString *TrueImageFadeKeyPathName(TrueImageFadeKeyPath keyPath)
{
  return keyPath == TrueImageFadeKeyPathContents ? @"contents" : @"opacity";
}

NSTimeInterval TrueImageFadeDuration(NSInteger transitionMs)
{
  return (NSTimeInterval)transitionMs / 1000.0;
}

NSURL *TrueImageURLFromSource(NSString *source)
{
  NSURLComponents *components = [NSURLComponents componentsWithString:source];
  if (components.scheme.length > 0) {
    return components.URL;
  }
  // NSURLComponents rejects unescaped spaces; retry with them encoded so a
  // `file:///…/My Cover.png` still parses. Anything without a scheme is a
  // resource name.
  NSString *escaped = [source stringByReplacingOccurrencesOfString:@" " withString:@"%20"];
  if ([escaped isEqualToString:source]) {
    return nil;
  }
  NSURLComponents *retry = [NSURLComponents componentsWithString:escaped];
  return retry.scheme.length > 0 ? retry.URL : nil;
}

File: ios/TrueImagePolicy.swift
import CoreGraphics
import Foundation

/// Pure geometry and policy shared by the view, the thumbnail tier and the
/// unit tests. Nothing here touches UIKit so it can run on any Swift host.
@objc public enum TrueImageFitMode: Int {
  case cover
  case contain
  case stretch
  case center
}

public enum TrueImagePolicy {
  /// The rect an image of `intrinsic` size occupies inside `bounds`.
  /// `cover` overflows, `contain` insets, both are centred. `stretch` fills
  /// the bounds, `center` draws at intrinsic size. A degenerate intrinsic
  /// size falls back to the full bounds.
  public static func fitRect(
    bounds: CGRect,
    intrinsic: CGSize,
    mode: TrueImageFitMode
  ) -> CGRect {
    guard intrinsic.width > 0, intrinsic.height > 0,
      bounds.width > 0, bounds.height > 0
    else { return bounds }

    let size: CGSize
    switch mode {
    case .stretch:
      return bounds
    case .center:
      size = intrinsic
    case .cover, .contain:
      let sx = bounds.width / intrinsic.width
      let sy = bounds.height / intrinsic.height
      let scale = mode == .cover ? max(sx, sy) : min(sx, sy)
      size = CGSize(width: intrinsic.width * scale, height: intrinsic.height * scale)
    }
    return CGRect(
      x: bounds.midX - size.width / 2,
      y: bounds.midY - size.height / 2,
      width: size.width,
      height: size.height
    )
  }

  /// Pixel size to resample a decoded image to when it is drawn far below
  /// its own size. Returns nil when no thumbnail is worth making: the view
  /// has no size, the mode does not scale, or the image is under twice the
  /// drawn size on either axis.
  public static func thumbnailPixelSize(
    imagePixels: CGSize,
    drawnPoints: CGSize,
    scale: CGFloat,
    mode: TrueImageFitMode
  ) -> CGSize? {
    guard drawnPoints.width > 0, drawnPoints.height > 0, scale > 0,
      imagePixels.width > 0, imagePixels.height > 0,
      mode != .center
    else { return nil }

    let drawnPixels = CGRect(
      origin: .zero,
      size: CGSize(width: drawnPoints.width * scale, height: drawnPoints.height * scale)
    )
    let target = fitRect(bounds: drawnPixels, intrinsic: imagePixels, mode: mode).size
    guard imagePixels.width >= target.width * 2, imagePixels.height >= target.height * 2
    else { return nil }
    return CGSize(width: ceil(target.width), height: ceil(target.height))
  }

  /// Memory cache ceiling: a sixteenth of physical memory, capped at 256 MB.
  /// Set because low-memory devices running long sessions were being
  /// terminated under the library default.
  public static func memoryCacheCost(physicalMemory: UInt64) -> UInt {
    let cap: UInt64 = 256 * 1024 * 1024
    return UInt(min(physicalMemory / 16, cap))
  }

  /// A fade softens an image arriving late; a cached image in an empty view
  /// is not late and draws at once. Replacing a displayed image always
  /// crossfades when a transition is set. Native resources never fade.
  public static func shouldFade(
    transitionMs: Int,
    fromMemory: Bool,
    hasContents: Bool,
    isResource: Bool
  ) -> Bool {
    guard transitionMs > 0, !isResource else { return false }
    return !(fromMemory && !hasContents)
  }

  public enum FadeKeyPath: String {
    case opacity
    case contents
  }

  /// An empty layer fades its opacity in; a layer that already shows an
  /// image crossfades its contents so nothing double-composites.
  public static func fadeKeyPath(hasContents: Bool) -> FadeKeyPath {
    hasContents ? .contents : .opacity
  }

  public static func fadeDuration(transitionMs: Int) -> TimeInterval {
    TimeInterval(transitionMs) / 1000
  }

  /// Turns a scheme-less name into nil so the caller takes the native
  /// resource branch; accepts `file://` paths containing spaces.
  public static func url(from source: String) -> URL? {
    guard let components = URLComponents(string: source), components.scheme != nil else {
      // URLComponents rejects unescaped spaces; retry with them encoded so a
      // `file:///…/My Cover.png` still parses. Anything without a scheme is a
      // resource name.
      let escaped = source.replacingOccurrences(of: " ", with: "%20")
      guard escaped != source,
        let retry = URLComponents(string: escaped), retry.scheme != nil
      else { return nil }
      return retry.url
    }
    return components.url
  }

  /// Cache key for a resampled thumbnail. Varies with the source, the blur
  /// and the target size so a resize or a blur change never reuses a stale one.
  public static func thumbnailKey(source: String, blurRadius: CGFloat, pixelSize: CGSize) -> String {
    "\(source)|b\(blurRadius)|\(Int(pixelSize.width))x\(Int(pixelSize.height))"
  }
}
